### ISSUE TITLE
Tracks in 3D showing in Z-stack and not time channel, also error when trying to save the tracks (Napari Plugin)

### ISSUE THREAD

**Issue by AKirschner88** [royerlab/ultrack#156]
Hi, 

When I important my video in to Napari and run the plugin, the conturs, foreground and segmentation all work and show good results, but when it finishes the tracks i get a problem.

the tracks move when I go through my Z-stack and not through the different timepoints, making it unable for me to verify the tacks visualy. Also, when I export the tracks I get an error. that I copied here below. 

The tracks look to be corrected if I see how the cells are moving, but I can not be certain.

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~\Anaconda3\envs\tracking\lib\site-packages\app_model\backends\qt\_qaction.py:57, in QCommandAction._on_triggered(self=QMenuItemAction(MenuItem(when=None, group='4_sav...tle=None, toggled=None), alt=None), app='napari'), checked=False)
     53 def _on_triggered(self, checked: bool) -> None:
     54     # execute_command returns a Future, for the sake of eventually being
     55     # asynchronous without breaking the API.  For now, we call result()
     56     # to raise any exceptions.
---> 57     self._app.commands.execute_command(self._command_id).result()
        self._command_id = 'napari.window.file.save_layers_dialog.selected'
        self = QMenuItemAction(MenuItem(when=None, group='4_save', order=None, command=CommandRule(id='napari.window.file.save_layers_dialog.selected', title='Save Selected Layers...', category=None, tooltip=None, status_tip=None, icon=None, icon_visible_in_menu=True, enablement=Expr.parse('num_selected_layers > 0'), short_title=None, toggled=None), alt=None), app='napari')
        self._app = Application('napari')

File ~\Anaconda3\envs\tracking\lib\site-packages\app_model\registries\_commands_reg.py:245, in CommandsRegistry.execute_command(self=<CommandsRegistry at 0x1def4dbb280 (146 commands)>, id='napari.window.file.save_layers_dialog.selected', execute_asynchronously=False, *args=(), **kwargs={})
    241 except Exception as e:
    242     if self._raise_synchronous_exceptions:
    243         # note, the caller of this function can also achieve this by
    244         # calling `future.result()` on the returned future object.
--> 245         raise e
    246     future.set_exception(e)
    248 return future

File ~\Anaconda3\envs\tracking\lib\site-packages\app_model\registries\_commands_reg.py:240, in CommandsRegistry.execute_command(self=<CommandsRegistry at 0x1def4dbb280 (146 commands)>, id='napari.window.file.save_layers_dialog.selected', execute_asynchronously=False, *args=(), **kwargs={})
    238 future: Future = Future()
    239 try:
--> 240     future.set_result(cmd(*args, **kwargs))
        future = <Future at 0x1df20734a00 state=pending>
        cmd = <function _save_selected_layers at 0x000001DF1CA8FF40>
        args = ()
        kwargs = {}
    241 except Exception as e:
    242     if self._raise_synchronous_exceptions:
    243         # note, the caller of this function can also achieve this by
    244         # calling `future.result()` on the returned future object.

File ~\Anaconda3\envs\tracking\lib\site-packages\in_n_out\_store.py:934, in Store.inject_processors.<locals>._deco.<locals>._exec(*args=(), **kwargs={})
    932 @wraps(func)
    933 def _exec(*args: P.args, **kwargs: P.kwargs) -> R:
--> 934     result = func(*args, **kwargs)
        func = <function _save_selected_layers at 0x000001DF317DAA70>
        args = ()
        kwargs = {}
    935     if result is not None:
    936         self.process(
    937             result,
    938             type_hint=type_hint,
   (...)
    941             _funcname=getattr(func, "__qualname__", str(func)),
    942         )

File ~\Anaconda3\envs\tracking\lib\site-packages\in_n_out\_store.py:804, in Store.inject.<locals>._inner.<locals>._exec(*args=(), **kwargs={})
    797 logger.debug(
    798     "  Calling %s with %r (injected %r)",
    799     _fname,
    800     bound.arguments,
    801     _injected_names,
    802 )
    803 try:
--> 804     result = func(**bound.arguments)
        bound = <BoundArguments (qt_viewer=<napari._qt.qt_viewer.QtViewer object at 0x000001DEF4E90430>)>
        func = <function _save_selected_layers at 0x000001DEF5415360>
        bound.arguments = {'qt_viewer': <napari._qt.qt_viewer.QtViewer object at 0x000001DEF4E90430>}
    805 except TypeError as e:
    806     if "missing" not in e.args[0]:

File ~\Anaconda3\envs\tracking\lib\site-packages\napari\_qt\_qapp_model\qactions\_file.py:105, in _save_selected_layers(qt_viewer=<napari._qt.qt_viewer.QtViewer object>)
    104 def _save_selected_layers(qt_viewer: QtViewer):
--> 105     qt_viewer._save_layers_dialog(selected=True)
        qt_viewer = <napari._qt.qt_viewer.QtViewer object at 0x000001DEF4E90430>

File ~\Anaconda3\envs\tracking\lib\site-packages\napari\_qt\qt_viewer.py:804, in QtViewer._save_layers_dialog(self=<napari._qt.qt_viewer.QtViewer object>, selected=True)
    800 writer = _npe2_decode_selected_filter(
    801     ext_str, selected_filter, writers
    802 )
    803 with warnings.catch_warnings(record=True) as wa:
--> 804     saved = self.viewer.layers.save(
        self = <napari._qt.qt_viewer.QtViewer object at 0x000001DEF4E90430>
        self.viewer.layers = [<Image layer 'Timelapse Video (w01)' at 0x1def81a79d0>, <Image layer 'Foreground' at 0x1df1d5aa8c0>, <Image layer 'Contours' at 0x1df1d5abdf0>, <Labels layer 'Segments' at 0x1df3171f250>, <Tracks layer 'Tracks' at 0x1df39ef94e0>]
        self.viewer = Viewer(camera=Camera(center=(0.0, 136.66428262078077, 151.58152497278513), zoom=4.983798537741507, angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True), cursor=Cursor(position=(0.0, 48.0, 252.13845408003593, 319.42539028986727), scaled=True, style=<CursorStyle.STANDARD: 'standard'>, size=1.0), dims=Dims(ndim=4, ndisplay=2, order=(0, 1, 2, 3), axis_labels=('0', '1', '2', '3'), rollable=(True, True, True, True), range=(RangeTuple(start=0.0, stop=23.0, step=1.0), RangeTuple(start=0.0, stop=62.0, step=2.0), RangeTuple(start=0.0, stop=255.5, step=0.5), RangeTuple(start=0.0, stop=255.5, step=0.5)), margin_left=(0.0, 0.0, 0.0, 0.0), margin_right=(0.0, 0.0, 0.0, 0.0), point=(0.0, 48.0, 127.5, 127.5), last_used=1), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[<Image layer 'Timelapse Video (w01)' at 0x1def81a79d0>, <Image layer 'Foreground' at 0x1df1d5aa8c0>, <Image layer 'Contours' at 0x1df1d5abdf0>, <Labels layer 'Segments' at 0x1df3171f250>, <Tracks layer 'Tracks' at 0x1df39ef94e0>], help='use <2> for transform', status={'layer_name': 'Tracks', 'layer_base': 'Tracks', 'source_type': '', 'plugin': '', 'coordinates': ' [48 252 319]'}, tooltip=Tooltip(visible=False, text=''), theme='dark', title='napari', mouse_over_canvas=True, mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x000001DE7B5189D0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, _keymap={})
        filename = 'C:/Users/Akirschner/Tracks'
        selected = True
        writer = None
    805         filename, selected=selected, _writer=writer
    806     )
    807     logging.debug('Saved %s', saved)
    808     error_messages = '\n'.join(str(x.message.args[0]) for x in wa)

File ~\Anaconda3\envs\tracking\lib\site-packages\napari\components\layerlist.py:486, in LayerList.save(self=[<Image layer 'Timelapse Video (w01)' at 0x1def8...1f250>, <Tracks layer 'Tracks' at 0x1df39ef94e0>], path='C:/Users/Akirschner/Tracks', selected=True, plugin=None, _writer=None)
    483     warnings.warn(msg)
    484     return []
--> 486 return save_layers(path, layers, plugin=plugin, _writer=_writer)
        layers = [<Tracks layer 'Tracks' at 0x1df39ef94e0>]
        path = 'C:/Users/Akirschner/Tracks'
        plugin = None
        _writer = None

File ~\Anaconda3\envs\tracking\lib\site-packages\napari\plugins\io.py:244, in save_layers(path='C:/Users/Akirschner/Tracks', layers=[<Tracks layer 'Tracks'>], plugin=None, _writer=None)
    240     written, writer_name = _write_multiple_layers_with_plugins(
    241         path, layers, plugin_name=plugin, _writer=_writer
    242     )
    243 elif len(layers) == 1:
--> 244     _written, writer_name = _write_single_layer_with_plugins(
        writer_name = ''
        layers = [<Tracks layer 'Tracks' at 0x1df39ef94e0>]
        path = 'C:/Users/Akirschner/Tracks'
        plugin = None
        _writer = None
        layers[0] = <Tracks layer 'Tracks' at 0x1df39ef94e0>
    245         path, layers[0], plugin_name=plugin, _writer=_writer
    246     )
    247     written = [_written] if _written else []
    248 else:

File ~\Anaconda3\envs\tracking\lib\site-packages\napari\plugins\io.py:464, in _write_single_layer_with_plugins(path='C:/Users/Akirschner/Tracks', layer=<Tracks layer 'Tracks'>, plugin_name=None, _writer=None)
    461     return (written_paths[0], writer_name)
    462 logger.debug('Falling back to original plugin engine.')
--> 464 hook_caller = getattr(
        plugin_manager.hook = <napari_plugin_engine.manager._HookRelay object at 0x000001DE7B5821A0>
        layer = <Tracks layer 'Tracks' at 0x1df39ef94e0>
        plugin_manager = <napari.plugins._plugin_manager.NapariPluginManager object at 0x000001DE7B580820>
    465     plugin_manager.hook, f'napari_write_{layer._type_string}'
    466 )
    468 if not plugin_name and isinstance(path, (str, pathlib.Path)):
    469     extension = os.path.splitext(path)[-1]

File ~\Anaconda3\envs\tracking\lib\site-packages\napari_plugin_engine\manager.py:903, in _HookRelay.__getattribute__(self=<napari_plugin_engine.manager._HookRelay object>, name='napari_write_tracks')
    901     if self._needs_discovery:
    902         self._manager.discover()
--> 903 return object.__getattribute__(self, name)
        name = 'napari_write_tracks'
        self = <napari_plugin_engine.manager._HookRelay object at 0x000001DE7B5821A0>

AttributeError: '_HookRelay' object has no attribute 'napari_write_tracks'```

**Comment by JoOkuma** [comment]
Hi @AKirschner88 could you share the dimensionality of your data and the resulting tracks?

You can get their shape with 
```python
viewer.layers[<REPLACE WITH LAYER NAME>].data.shape
```

In the napari console, lower left menu. Or in your own jupyter notebook.

**Comment by JoOkuma** [comment]
@AKirschner88, I found the bug, I opened #158 and it will be fixed ASAP. cc @ilan-theodoro

**Comment by JoOkuma** [comment]
closed in #164
